README.md
Эта библиотека позволяет "общаться" торговым терминалам [**Evraon**](https://github.com/avently/Evraon) и **Quik**.
Через нее можно получать информацию с бирж и отправлять заявки. Цель проекта проста: дать возможность 
пользователям Evraon торговать на Московской и Санкт-Петербургской биржах, а никак не повторение всего API Quik'a. 
То есть реализовано лишь то, что требуется в данный момент. Заявки на новый функционал не принимается, если это не касается 
непосредственно Эвраона.


Зачем эти скрипты нужны
-----------------------

Для начала неплохо бы понять, зачем они вообще нужны. К сожалению, нельзя торговать на биржах в обход Квика. И вот для того, чтобы 
иметь таки возможность торговать, придуман Lua API Квика. Это некая договоренность двух программ, как обмениваться информацией. 
Эти скрипты запускают два сокет сервера, к которым подключается программа, и начинается взаимная отправка и получение информации.
При этом запросы от программы передаются непосредственно в Квик.

Как использовать
----------------

Скачайте на [этой странице](https://github.com/avently/EvraonQ/releases) архив с версией, равной версии Evraon. 
Если таковой нет, а у вас самая новая версия Evraon, то скачивайте самый верхний архив. Распакуйте его в любую папку.
Затем зайдите в Квик, выберите меню `Сервисы` -> `Lua скрипты...`, затем `Добавить`, выберите файл под названием `Server.lua`,
`Открыть`, нажмите `Запустить`. Все, теперь можно подключаться к серверу из программы Evraon, а вот и 
[инструкция](https://evraon.com/ru/docs/connectors/quik), как это сделать.


Как подключиться к двум терминалам сразу
----------------------------------------

Достаточно сделать ряд простых действий:
1. Найдите в файле `config.json` секцию `servers`, там будет название опции `scriptName` со значением `Server_2`. В данном случае 
`Server_2` является названием файла в папке lua (расширение `.lua` в `config.json` указывать нельзя).
2. Измените опции согласно вашим предпочтениям. По умолчанию, дополнительный сокет сервер будет запущен на localhost'e (т.е. опции 
`responseHostname` и `callbackHostname` имеют значение `127.0.0.1`), а порты (опции `responsePort` и `callbackPort`) равны `15515` 
для сервера запросов-ответов и `15516` для сервера коллбэков (данные, которые хочет отправить Квик). 
Порты не должны конфликтовать с уже указанными портами в файле `config.json`
3. Добавьте `Server_2.lua` в Quik и запустите. При этом необязательно запускать `Server.lua`, 
если он вами не используется. Если используется и для него указаны другие порты, то, безусловно, можно запускать оба.

Все!

Вопросы и ответы
----------------

- Как добавить еще сервера?

Скопируйте файл `Server_2.lua` и сохраните его в той же папке под другим именем, например, `MyBroker.lua`. 
В файле `config.json` в секции `servers` скопируйте текст внутри фигурных скобок 
`{"scriptName"..."responseHostname"..."responsePort"..."callbackHostname"..."callbackPort"...}`, поставьте запятую после `}`, 
вставьте скопированный текст и измените в нем данные. Как минимум, нужно указать в `scriptName` название файла скрипта 
(в примере это `MyBroker`), а в опциях портов указать любые свободные номера портов. 
Такую процедуру можно делать столько раз, сколько серверов вам нужно. 
Не забудьте, что `responsePort` и `callbackPort` не могут быть одинаковыми и повторяться в файле.

- Как обновляться на новые версии библиотеки и скриптов в частности?

В новых версиях будет меняться код скриптов и код `config.json`. Но если вы вносили изменения в `config.json`, 
то сохраните в другом месте файл с вашими изменениями. После этого замените все старые файлы новыми. 
Обратите внимание, что, возможно, в одном из обновлений содержимое `Server_2.lua` изменится и все созданные вами лично копии 
этого файла потребуется пересоздать на основе обновленного файла. 
Продолжаем. Отредактируйте файл `config.json` в соответствии с настройками из старого файла 
(если не меняли ничего, делать ничего не надо). Процесс обновления завершен.

- Как сделать так, чтобы можно было подключаться к сокету из локальной сети WiFi с другого устройства?

Нужно указать в `responseHostname` и `callbackHostname` значение `0.0.0.0`. Лучше так не делать и вот, почему. 
Когда компьютер подключен к роутеру, то разрешение подключаться к нему извне (а именно это делает `0.0.0.0`) 
открывает доступ к нему всему миру при допущении, что роутер недостаточно хорошо защищен и настроен. 
А так как при подключении к сокетам нет ни пароля, ни шифрования, открытый доступ всему миру открывает ваш Квик всем. 
Т.е. если вы все же хотите подключаться из локальной сети, позаботьтесь о безопасности и верных настройках роутера.

- Как можно подключаться из локальной сети безопасно?

Просто подключайтесь к серверу Evraon, запущенному на вашем компьютере. `responseHostname` и `callbackHostname` должны быть равны 
`127.0.0.1`. В таком случае подключение будет защищено логином и паролем, а само соединение между мобильной версией 
и настольной будет зашифровано. 

- Можно ли переименовать `Server_2.lua`?

Да, только укажите новое имя в `scriptName` файла `config.json`.


- Можно ли переименовать `Server.lua`?

Нет. Этот файл обязательно должен сохранить свое имя, потому что все остальные скрипты используют часть его кода.

- Создается папка `logs`, зачем она?

Там пишется диагностическая информация и ошибки. Иногда полезно туда заглядывать в поиске чего-то необычного.

CHANGELOG.md
# Список изменений

## v1.0.0

* Первый релиз

Server.lua
pathSeparator = package.config:sub(1,1)
local scriptPath
local linuxLibDir = ""
if pathSeparator == '/' then
    linuxLibDir = "/usr/lib/lua/5.3"
end
if getScriptPath then
    scriptPath = getScriptPath() .. "\\"
else
    utils = require "utils"
    scriptPath = utils.script_path()
end
local libDir
if not getScriptPath or tonumber(string.match(getInfoParam("VERSION"), "%d+[.]%d+")) >= 8.5 then
    libDir = "53"
elseif tonumber(string.match(getInfoParam("VERSION"), "%d+[.]%d+")) >= 8 then
    libDir = "51-x64"
else
    libDir = "51-x86"
end

package.path = package.path .. ";" .. scriptPath .. "?.lua"
package.cpath = package.cpath .. ";".. linuxLibDir .. scriptPath .. "libs".. pathSeparator .. libDir .. pathSeparator .. "?.dll"

socketsAttached = false
socket = require "libs/socket"
json = require "libs/dkjson"
utils = require "utils"
routes = require "routes"
require "callbacks"

function main()
    setup(utils.script_file_name())
end

function setup(scriptName)
    if not scriptName then
        --notify("Не удалось автоматически определить название файла. Пожалуйста, напишите его сами вместо script_file_name() " ..
        --        "в файле, который запущен в Квике")
        notify("File name of this script is unknown. Please, set it explicitly instead of script_file_name() " ..
                "call inside your custom file")
        return false
    end

    local list = utils.params_from_config(scriptName)
    if list and #list == 4 then
        local responseHost, responsePort, callbackHost, callbackPort = unpack(list)
        utils.print_running_message(responseHost, responsePort, callbackHost, callbackPort)
        start_waiting(responseHost, responsePort, callbackHost, callbackPort)
    elseif scriptName == "Server" then
        -- use default values for this file in case no custom config found for it
        local responseHost, responsePort, callbackHost, callbackPort = { "127.0.0.1", 15515, "127.0.0.1", 15516 }
        utils.print_running_message(responseHost, responsePort, callbackHost, callbackPort)
        start_waiting(responseHost, responsePort, callbackHost, callbackPort)
    else
        -- do nothing when config is not found
        --notify("Файл config.json не найден или не содержит записи о скрипте: " .. scriptName .. ".lua")
        notify("File config.json is not found or contains no entries for this script: " .. scriptName .. ".lua")
        return false
    end

    return true
end

local mainSocket
local callbackSocket

function start_waiting(responseHost, responsePort, callbackHost, callbackPort)
    --notify(responseHost .. responsePort .. callbackHost .. callbackPort)
    local s1 = assert(socket.bind(responseHost, responsePort))
    local s2 = assert(socket.bind(callbackHost, callbackPort))
    while true and s1 and s2 do
        mainSocket = assert(s1:accept())
        callbackSocket = assert(s2:accept())
        --notify("Evraon соединился")
        notify("Evraon: Connected")
        socketsAttached = true
        local req, error = mainSocket:receive()
        while not error do
            -- print("REQ " .. req)
            local parsedReq = json.decode(req)
            if type(parsedReq) == "table" then
                process_request(parsedReq)
            else
                -- Received something that is not a table, close connection, unexpected situation
                break
            end
            mainSocket:send(json.encode(parsedReq) .. '\n')
            req, error = mainSocket:receive()
        end
        socketsAttached = false
        pcall(callbackSocket.send, callbackSocket, "null")
        pcall(mainSocket.close, mainSocket)
        pcall(callbackSocket.close, callbackSocket)
        mainSocket = nil
        callbackSocket = nil
        --notify("Evraon отсоединился")
        notify("Evraon: Disconnected")
        socket.sleep(1)
    end
    s1:close()
    s2:close()
end

function process_request(request)
    local statusOk, res, errFromRoute = pcall(route_command, request.c, request.p)
    request.p = nil
    if statusOk then
        if not errFromRoute then
            request.r = res
        else
            request.e = errFromRoute
        end
    else
        request.e = res
    end
end

function route_command(command, params)
    action = {
        ["get_security_info"] = function() return routes.get_security_info(split(params)) end,
        ["get_class_securities"] = function() return routes.get_class_securities(params) end,
        ["get_class_info"] = function() return routes.get_class_info(params) end,
        ["get_info_param"] = function() return routes.get_info_param(params) end,
        ["get_futures_holding"] = function() return routes.get_futures_holding(split(params)) end,
        ["get_futures_client_limits"] = function() return routes.get_futures_client_limits() end,
        ["get_depo_limit"] = function() return routes.get_depo_limit(split(params)) end,
        ["get_money_limits"] = function() return routes.get_money_limits() end,
        ["get_trade_accounts"] = function() return routes.get_trade_accounts() end,
        ["get_candlesticks"] = function() return routes.get_candlesticks(split(params)) end,
        ["get_quote_level_2"] = function() return routes.get_quote_level_2(split(params)) end,

        ["subscribe_level_2_quotes"] = function() return routes.subscribe_level_2_quotes(split(params)) end,
        ["unsubscribe_level_2_quotes"] = function() return routes.unsubscribe_level_2_quotes(split(params)) end,
        ["is_subscribed_level_2_quotes"] = function() return routes.is_subscribed_level_2_quotes(split(params)) end,

        ["get_security_info_bulk"] = function() return routes.get_security_info_bulk(params) end,
        ["get_param_ex2_bulk"] = function() return routes.get_param_ex2_bulk(params) end,
        ["param_request_bulk"] = function() return routes.param_request_bulk(params) end,
        ["cancel_param_request_bulk"] = function() return routes.cancel_param_request_bulk(params) end,

        ["get_orders"] = function() return routes.get_orders(split(params)) end,
        ["get_order_by_id"] = function() return routes.get_order_by_id(split(params)) end,
        ["get_order_by_num"] = function() return routes.get_order_by_num(params) end,
        ["get_stop_orders"] = function() return routes.get_stop_orders(split(params)) end,
        ["get_stop_order_by_id"] = function() return routes.get_stop_order_by_id(split(params)) end,
        ["get_stop_order_by_num"] = function() return routes.get_stop_order_by_num(params) end,
        ["send_transaction"] = function() return routes.send_transaction(params) end,

    }
    local act = action[command]
    if act then
        return action[command]()
    else
        return nil, "No such function"
    end
end

function new_callback(command, data)
    --log_i(command, data)
    if callbackSocket then
        local res = {}
        res.c = command
        res.r = data
        local statusOk, err = pcall(callbackSocket.send, callbackSocket, json.encode(res) .. '\n')
        if not statusOk then
            --notify("Не удалось отправить коллбэк: " .. err)
            notify("Unable to send a callback: " .. err)
            --log_e("Коллбэк не был отправлен, отсоединяюсь: " .. err)
            log_e("Callback wasn't sent, disconnecting: " .. err)
            if mainSocket then
                pcall(mainSocket.close, mainSocket)
            end
        end
    end
end

function on_stop()
    if mainSocket then
        pcall(mainSocket.close, mainSocket)
    end
    --log_i("Работа прекращена")
    log_i("Stopped working")
    utils.close_log_file()
end

--if getScriptPath == nil and utils.script_file_name() == "Server" then main() end

routes.lua
local json = require("libs/dkjson")
local routes = {}

function routes.get_security_info(secCode, classCode)
    return getSecurityInfo(classCode, secCode)
end

function routes.get_class_info(classCode)
    return getClassInfo(classCode)
end

function routes.get_class_securities(classCode)
    local res = split_in_table(getClassSecurities(classCode), ",")
    if #res >= 1 then
        res[#res] = nil
    end
    return res
end

function routes.get_info_param(paramName)
    return getInfoParam(paramName)
end

function routes.get_futures_holding(firmId, trdAccId, secCode, type)
    return getFuturesHolding(firmId, trdAccId, secCode, tonumber(type))
end

function routes.get_futures_client_limits()
    local res = {}
    for i = 0, getNumberOf("futures_client_limits") - 1 do
        table.insert(res, getItem("futures_client_limits", i))
    end
    return res
end

function routes.get_depo_limit(firmId, clientCode, secCode, trdAccId, limitKind)
    return getDepoEx(firmId, clientCode, secCode, trdAccId, tonumber(limitKind))
end

function routes.get_money_limits()
    local res = {}
    for i = 0, getNumberOf("money_limits") - 1 do
        table.insert(res, getItem("money_limits", i))
    end
    return res
end

function routes.get_trade_accounts()
    local res = {}
    for i = 0, getNumberOf("trade_accounts") - 1 do
        local account = getItem("trade_accounts", i)
        if account.class_codes ~= "" then
            table.insert(res, account)
        end
    end
    return res
end

function routes.get_candlesticks(secCode, classCode, interval, limit, skip)
    --log_i("Candles loading started " .. secCode .. " class " .. classCode .. " interval " .. interval)
    local dataSource, err = CreateDataSource(classCode, secCode, tonumber(interval))
    local limit = tonumber(limit)
    if limit == 0 then
        limit = 10000000
    end
    local skip = tonumber(skip)
    if dataSource then
        local res = {}
        local sleepTime = 0
        while dataSource:Size() == 0 and sleepTime < 5000 do
            sleep(10)
            sleepTime = sleepTime + 10
        end
        --log_i("Candles loading ended " .. sleepTime)
        local size = dataSource:Size()
        if size == 0 then
            return nil
        end

        local start = math.max(1, size - limit - skip)
        local to = math.min(size, start - 1 + limit)
        for i = start - 1, to do
            local candle = {}
            candle.O = dataSource:O(i)
            candle.H = dataSource:H(i)
            candle.L = dataSource:L(i)
            candle.C = dataSource:C(i)
            candle.V = dataSource:V(i)
            candle.T = dataSource:T(i)
            table.insert(res, candle)
        end
        dataSource:Close()
        return res
    else
        log_e("Candlesticks were not loaded: " .. err)
        return nil, err
    end
end

function routes.get_quote_level_2(secCode, classCode)
    return getQuoteLevel2(classCode, secCode)
end

function routes.subscribe_level_2_quotes(secCode, classCode)
    return Subscribe_Level_II_Quotes(classCode, secCode)
end

function routes.unsubscribe_level_2_quotes(secCode, classCode)
    return Unsubscribe_Level_II_Quotes(classCode, secCode)
end

function routes.is_subscribed_level_2_quotes(secCode, classCode)
    return IsSubscribed_Level_II_Quotes(classCode, secCode)
end

--
-- ORDERS
--

function routes.get_orders(secCode, classCode)
    local res = {}
    for i = 0, getNumberOf("orders") - 1 do
        local order = getItem("orders", i)
        if order.sec_code == secCode and order.class_code == classCode then
            table.insert(res, routes.fill_order_with_trades_data(order))
        end
    end
    return res
end

function routes.get_order_by_id(secCode, classCode, transId)
    local transId = tonumber(transId)
    for i = getNumberOf("orders") - 1, 0, -1 do
        local order = getItem("orders", i)
        if order.sec_code == secCode and order.class_code == classCode and order.trans_id == transId then
            return routes.fill_order_with_trades_data(order)
        end
    end
    return nil
end

function routes.get_order_by_num(orderNumber)
    local orderNum = tonumber(orderNumber)
    for i = getNumberOf("orders") - 1, 0, -1 do
        local order = getItem("orders", i)
        if order.order_num == orderNum then
            return routes.fill_order_with_trades_data(order)
        end
    end
    return nil
end

function routes.get_stop_orders(secCode, classCode)
    local res = {}
    for i = 0, getNumberOf("stop_orders") - 1 do
        local order = getItem("stop_orders", i)
        if order.sec_code == secCode and order.class_code == classCode then
            table.insert(res, order)
        end
    end
    return res
end

function routes.get_stop_order_by_id(secCode, classCode, transId)
    local transId = tonumber(transId)
    for i = getNumberOf("stop_orders") - 1, 0, -1 do
        local order = getItem("stop_orders", i)
        if order.sec_code == secCode and order.class_code == classCode and order.trans_id == transId then
            return routes.fill_order_with_trades_data(order)
        end
    end
    return nil
end

function routes.get_stop_order_by_num(orderNumber)
    local orderNum = tonumber(orderNumber)
    for i = getNumberOf("stop_orders") - 1, 0, -1 do
        local order = getItem("stop_orders", i)
        if order.order_num == orderNum then
            return routes.fill_order_with_trades_data(order)
        end
    end
    return nil
end

function routes.get_trades_by_num(orderNumber)
    local orderNum = tonumber(orderNumber)
    local res = {}
    for i = 0, getNumberOf("trades") - 1 do
        local trade = getItem("trades", i)
        if trade.order_num == orderNum then
            table.insert(res, trade)
        end
    end
    return res
end

function routes.fill_order_with_trades_data(order)
    local qty_of_trades = 0
    local price_qty = 0
    for i = 0, getNumberOf("trades") - 1 do
        local trade = getItem("trades", i)
        if trade.order_num == order.order_num then
            price_qty = price_qty + trade.price * trade.qty
            qty_of_trades = qty_of_trades + trade.qty
            order.last_trade_datetime_ex = trade.datetime
        end
    end
    if qty_of_trades > 0 then
        order.average_price_ex = price_qty / qty_of_trades
    end
    return order
end

function routes.send_transaction(transaction)
    local err = sendTransaction(transaction)
    if err ~= "" then
        return nil, err
    else
        return true
    end
end

--
-- BULK REQUESTS
--

function routes.get_security_info_bulk(secsClasses)
    local res = {}
    for i = 1, #secsClasses do
        local secCode, classCode = split(secsClasses[i])
        local security = getSecurityInfo(classCode, secCode)
        if security then
            table.insert(res, security)
        else
            table.insert(res, json.null)
        end
    end
    return res
end

function routes.get_param_ex2_bulk(secsClassesParams)
    local res = {}
    for i = 1, #secsClassesParams do
        local secCode, classCode, param = split(secsClassesParams[i])
        table.insert(res, getParamEx2(classCode, secCode, param))
    end
    return res
end

function routes.param_request_bulk(secsClassesParams)
    local res = {}
    for i = 1, #secsClassesParams do
        local secCode, classCode, param = split(secsClassesParams[i])
        table.insert(res, ParamRequest(classCode, secCode, param))
    end
    return res
end

function routes.cancel_param_request_bulk(secsClassesParams)
    local res = {}
    for i = 1, #secsClassesParams do
        local secCode, classCode, param = split(secsClassesParams[i])
        table.insert(res, CancelParamRequest(classCode, secCode, param))
    end
    return res
end

return routes

callbacks.lua
local callbacks = {}

function OnAccountBalance(acc_bal) new_callback("OnAccountBalance", acc_bal) end

function OnAccountPosition(acc_pos) new_callback("OnAccountPosition", acc_pos) end

function OnAllTrade(alltrade) new_callback("OnAllTrade", alltrade) end

function OnCleanUp() new_callback("OnCleanUp", nil) end

function OnClose() new_callback("OnClose", nil) end

function OnConnected(flag) new_callback("OnConnected", flag) end

function OnDepoLimit(dlimit) new_callback("OnDepoLimit", dlimit) end

function OnDepoLimitDelete(dlimit_del) new_callback("OnDepoLimitDelete", dlimit_del) end

function OnDisconnected() new_callback("OnDisconnected", flag) end

function OnFirm(firm) new_callback("OnFirm", firm) end

function OnFuturesClientHolding(fut_pos) new_callback("OnFuturesClientHolding", fut_pos) end

function OnFuturesLimitChange(fut_limit) new_callback("OnFuturesLimitChange", fut_limit) end

function OnFuturesLimitDelete(lim_del) new_callback("OnFuturesLimitDelete", lim_del) end

function OnInit(script_path) new_callback("OnInit", script_path) end

function OnMoneyLimit(mlimit) new_callback("OnMoneyLimit", mlimit) end

function OnMoneyLimitDelete(mlimit_del) new_callback("OnMoneyLimitDelete", mlimit_del) end

function OnNegDeal(neg_deals) new_callback("OnNegDeal", neg_deals) end

function OnNegTrade(neg_trade) new_callback("OnNegTrade", neg_trade) end

function OnOrder(order) new_callback("OnOrder", order) end

function OnParam(class_code, sec_code)
    local r = {}
    r.sec_code = sec_code
    r.class_code = class_code
    new_callback("OnParam", r)
end

function OnQuote(class_code, sec_code)
    if not socketsAttached then return end
    local statusOk, r = pcall(getQuoteLevel2, class_code, sec_code)
    if statusOk then
        r.sec_code = sec_code
        r.class_code = class_code
        new_callback("OnQuote", r)
    end
end

function OnStop(signal)
    new_callback("OnStop", signal)
    on_stop()
    return 500
end

function OnStopOrder(stop_order) new_callback("OnStopOrder", stop_order) end

function OnTrade(trade) new_callback("OnTrade", trade) end

function OnTransReply(trans_reply) new_callback("OnTransReply", trans_reply) end

return callbacks

utils.lua
local utils = {}

local delimiter = "|"
local cachedScriptName
local cachedScriptPath

-- GLOBAL
-- FUNCTIONS

function split(s)
    local res = {}
    for match in (s..delimiter):gmatch("(.-)"..delimiter) do
        table.insert(res, match);
    end
    return unpack(res);
end

function split_in_table(s, separator)
    local res = {}
    for match in (s..separator):gmatch("(.-)"..separator) do
        table.insert(res, match);
    end
    return res
end

function log_e(error)
    --utils.log("ОШИБКА  ".. utils.script_file_name() ..":  " .. error)
    utils.log("ERROR  ".. utils.script_file_name() ..":  " .. error)
end

function log_i(info)
    --utils.log("ИНФОРМ.  ".. utils.script_file_name() ..":  " .. info)
    utils.log("INFO  ".. utils.script_file_name() ..":  " .. info)
end

function notify(text)
    if getScriptPath then
        message(text)
    end
    log_i(text)
end


-- LOCAL
-- FUNCTIONS

-- Returns the name of the file that calls this function (without extension)
function utils.script_file_name()
    if cachedScriptName then
        return cachedScriptName
    end
    --message(debug.traceback("Trace"))
    -- Check that Lua runtime was built with debug information enabled
    if not debug or not debug.getinfo then
        return nil
    end
    local fullPath = debug.getinfo(2, "S").source:sub(2)
    cachedScriptName = string.gsub(fullPath, "^.*[\\/](.*)[.]lua[c]?$", "%1")
    return cachedScriptName
end

-- Returns the path of the file with separator
function utils.script_path()
    if cachedScriptPath ~= nil then
        return cachedScriptPath
    end

    -- Quik is running, return path from it
    if getScriptPath ~= nil then
        cachedScriptPath = getScriptPath() .. pathSeparator
        return cachedScriptPath
    end

    -- Check that Lua runtime was built with debug information enabled
    if not debug or not debug.getinfo then
        return nil
    end
    cachedScriptPath = string.gsub(debug.getinfo(1).source, "^@(.+[\\/])[^\\/]+$", "%1")
    return cachedScriptPath
end

-- Create log directory
function utils.get_log_file()
    local logsPath = utils.script_path() .. "logs"
    os.execute("mkdir \"".. logsPath .. "\"")
    -- Opens a file in append mode
    return io.open(logsPath .. pathSeparator .. os.date("%Y%m%d") .. ".txt", "a")
end

local log = utils.get_log_file()

function utils.close_log_file()
    log:close()
end

function utils.log(text)
    -- appends something to the last line of the file
    pcall(log.write, log, os.date("%d.%m.%Y, %X") .. "  " .. text .. '\n')
    pcall(log.flush, log)
end

-- Returns contents of config.json file or nil if no such file exists
function utils.read_config_as_json()
    local conf = io.open (utils.script_path() .. "config.json", "r")
    if not conf then
        return nil
    end
    local content = conf:read "*a"
    conf:close()
    return json.decode(content)
end

function utils.params_from_config(scriptName)
    local params = {}
    -- just default values
    table.insert(params, "127.0.0.1") -- responseHostname
    table.insert(params, 15515)       -- responsePort
    table.insert(params, "127.0.0.1") -- callbackHostname
    table.insert(params, 15516)       -- callbackPort

    local config = utils.read_config_as_json()
    if not config or not config.servers then
        return nil
    end
    local found = false
    for i=1,#config.servers do
        local server = config.servers[i]
        if server.scriptName == scriptName then
            found = true
            if server.responseHostname then
                params[1] = server.responseHostname
            end
            if server.responsePort then
                params[2] = server.responsePort
            end
            if server.callbackHostname then
                params[3] = server.callbackHostname
            end
            if server.callbackPort then
                params[4] = server.callbackPort
            end
        end
    end

    if found then
        return params
    else
        return nil
    end
end

function utils.print_running_message(responseHost, responsePort, callbackHost, callbackPort)
    --log_i("Запущен с параметрами: основной сокет " .. responseHost .. ":" .. responsePort ..", коллбэк ".. " "..
    --        callbackHost ..":".. callbackPort)
    log_i("Running with params: response " .. responseHost .. ":" .. responsePort ..", callback ".. " "..
            callbackHost ..":".. callbackPort)
end

return utils